Upgrading to 1.6.0-beta fails restoring Raft snapshots

### Nomad version
Nomad v1.6.0-beta.1+ent
BuildDate 2023-06-27T16:32:15Z
Revision b26558fbf8f4cf89dbaa29068620241746c583bb

(Upgrade from 1.5.6)

### Operating system and Environment details

Ubuntu 22.04, 3 nodes being server+client

### Issue

Doing an in-place upgrade (run Ansible to replace the binary with the 1.6.0-beta release, then started restarting the Nomad service one by one). The service fails with the following log:

```
Jun 29 10:05:44 nuc-02 systemd[1]: nomad.service: Failed with result 'exit-code'.
Jun 29 10:06:26 nuc-02 systemd[1]: nomad.service: Scheduled restart job, restart counter is at 1.
Jun 29 10:06:26 nuc-02 systemd[1]: Stopped nomad agent.
Jun 29 10:06:26 nuc-02 systemd[1]: Started nomad agent.
Jun 29 10:06:26 nuc-02 nomad[298802]: ==> WARNING: mTLS is not configured - Nomad is not secure without mTLS!
Jun 29 10:06:26 nuc-02 nomad[298802]: ==> Config enable_syslog is `true` with log_level=INFO
Jun 29 10:06:26 nuc-02 nomad[298802]: ==> Loaded configuration from /etc/nomad.d/base.hcl, /etc/nomad.d/client.hcl, /etc/nomad.d/server.hcl
Jun 29 10:06:26 nuc-02 nomad[298802]: ==> Starting Nomad agent...
Jun 29 10:06:26 nuc-02 nomad[298802]:  nomad: setting up raft bolt store: no_freelist_sync=false
Jun 29 10:06:26 nuc-02 nomad[298802]:  nomad.raft: starting restore from snapshot: id=3-8205-1687786395074 last-index=8205 last-term=3 size-in-bytes=83425
Jun 29 10:06:26 nuc-02 nomad[298802]:  nomad.raft: snapshot restore progress: id=3-8205-1687786395074 last-index=8205 last-term=3 size-in-bytes=83425 read-bytes=34130 percent-complete="40.91%"
Jun 29 10:06:26 nuc-02 nomad[298802]: nomad.raft: failed to restore snapshot: id=3-8205-1687786395074 last-index=8205 last-term=3 size-in-bytes=83425 error="job insert failed: missing value for index 'pool'"
Jun 29 10:06:26 nuc-02 nomad[298802]:  nomad: shutting down server
Jun 29 10:06:26 nuc-02 nomad[298802]: ==> Error starting agent: server setup failed: Failed to start Raft: failed to load any existing snapshots
Jun 29 10:06:26 nuc-02 nomad[298802]:     2023-06-29T10:06:26.460Z [INFO]  nomad: setting up raft bolt store: no_freelist_sync=false
Jun 29 10:06:26 nuc-02 nomad[298802]:     2023-06-29T10:06:26.462Z [INFO]  nomad.raft: starting restore from snapshot: id=3-8205-1687786395074 last-index=8205 last-term=3 size-in-bytes=83425
Jun 29 10:06:26 nuc-02 nomad[298802]:     2023-06-29T10:06:26.464Z [INFO]  nomad.raft: snapshot restore progress: id=3-8205-1687786395074 last-index=8205 last-term=3 size-in-bytes=83425 read-bytes=34130 percent-c>
Jun 29 10:06:26 nuc-02 nomad[298802]:     2023-06-29T10:06:26.464Z [ERROR] nomad.raft: failed to restore snapshot: id=3-8205-1687786395074 last-index=8205 last-term=3 size-in-bytes=83425 error="job insert failed:>
Jun 29 10:06:26 nuc-02 nomad[298802]:     2023-06-29T10:06:26.464Z [INFO]  nomad: shutting down server
Jun 29 10:06:26 nuc-02 nomad[298802]:     2023-06-29T10:06:26.464Z [ERROR] nomad: failed to start Raft: error="failed to load any existing snapshots"
Jun 29 10:06:26 nuc-02 nomad[298802]:     2023-06-29T10:06:26.464Z [ERROR] agent: error starting agent: error="server setup failed: Failed to start Raft: failed to load any existing snapshots"
Jun 29 10:06:26 nuc-02 nomad[298802]: nomad: failed to start Raft: error="failed to load any existing snapshots"
Jun 29 10:06:26 nuc-02 nomad[298802]: agent: error starting agent: error="server setup failed: Failed to start Raft: failed to load any existing snapshots"
Jun 29 10:06:26 nuc-02 systemd[1]: nomad.service: Main process exited, code=exited, status=1/FAILURE
Jun 29 10:06:26 nuc-02 systemd[1]: nomad.service: Failed with result 'exit-code'.
```


### Reproduction steps
Upgrade Nomad Ent from 1.5.6 to 1.6.0-beta

#### Expected Result

The upgrade suceeds.

#### Actual Result

Raft snapshots cannot be restored, *maybe* due to Node pools.

--- a/nomad/state/state_store_restore.go
+++ b/nomad/state/state_store_restore.go
@@ -44,6 +44,17 @@ func (r *StateRestore) NodePoolRestore(pool *structs.NodePool) error {
 
 // JobRestore is used to restore a job
 func (r *StateRestore) JobRestore(job *structs.Job) error {
+
+	// When upgrading a cluster pre to post 1.6, the existing jobs will not
+	// have a node pool set. Inserting this into the table will fail, as this
+	// is indexed and cannot be empty.
+	//
+	// This cannot happen within the job canonicalize function, as it would
+	// break the node pools and governance feature.
+	if job.NodePool == "" {
+		job.NodePool = structs.NodePoolDefault
+	}
+
 	if err := r.txn.Insert("jobs", job); err != nil {
 		return fmt.Errorf("job insert failed: %v", err)
 	}

--- a/nomad/state/state_store_restore_test.go
+++ b/nomad/state/state_store_restore_test.go
@@ -50,28 +50,36 @@ func TestStateStore_RestoreJob(t *testing.T) {
 	ci.Parallel(t)
 
 	state := testStateStore(t)
-	job := mock.Job()
+	mockJob1 := mock.Job()
 
 	restore, err := state.Restore()
-	if err != nil {
-		t.Fatalf("err: %v", err)
-	}
+	must.NoError(t, err)
 
-	err = restore.JobRestore(job)
-	if err != nil {
-		t.Fatalf("err: %v", err)
-	}
-	require.NoError(t, restore.Commit())
+	err = restore.JobRestore(mockJob1)
+	must.NoError(t, err)
+	must.NoError(t, restore.Commit())
 
 	ws := memdb.NewWatchSet()
-	out, err := state.JobByID(ws, job.Namespace, job.ID)
-	if err != nil {
-		t.Fatalf("err: %v", err)
-	}
+	out, err := state.JobByID(ws, mockJob1.Namespace, mockJob1.ID)
+	must.NoError(t, err)
+	must.Eq(t, mockJob1, out)
 
-	if !reflect.DeepEqual(out, job) {
-		t.Fatalf("Bad: %#v %#v", out, job)
-	}
+	// Test upgrade to 1.6 or greater to simulate restoring a job which does
+	// not have a node pool set.
+	mockJob2 := mock.Job()
+	mockJob2.NodePool = ""
+
+	restore, err = state.Restore()
+	must.NoError(t, err)
+
+	err = restore.JobRestore(mockJob2)
+	must.NoError(t, err)
+	must.NoError(t, restore.Commit())
+
+	ws = memdb.NewWatchSet()
+	out, err = state.JobByID(ws, mockJob2.Namespace, mockJob2.ID)
+	must.NoError(t, err)
+	must.Eq(t, structs.NodePoolDefault, out.NodePool)
 }
 
 func TestStateStore_RestorePeriodicLaunch(t *testing.T) {

